### PR TITLE
Service status and errors docs

### PR DESCRIPTION
**Pull request by FalkoBasner**
Start a discussion about service error situations with its corresponding user facing status codes and caching options.

### CODE REVIEW

File: docs/service_status_and_errors.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+https://grpc.github.io/grpc/core/md_doc_statuscodes.html
+
+## Article Page Service Status
+
+| Error scenario | GRPC status | HTTP Status | Cache Headers | Description |
+| --- |:---:| ---:| ---:| --- |
+| Article found | OK | 200 | max-age=60
+| Article not found | NOT_FOUND | 404 | max-age=60
+| Article not published yet | NOT_FOUND | 404 | max-age=60
+| Article expired | NOT_FOUND | 410 | max-age=60
```
[the remark this answers is not in the document]
Comment (moritzzimmer) [reply docs/service_status_and_errors.md:16]: Google proposes a similar mapping with recommendations for error details in grpc responses: https://cloud.google.com/apis/design/errors#error_payloads - we could use this to distinguish reasons of certain grpc errors

[the remark this answers is not in the document]
Comment (bobaaaaa) [reply docs/service_status_and_errors.md:16]: `Fehlernutzlasten` :D

**Comment by bobaaaaa** [comment]
What about redirects?

File: docs/service_status_and_errors.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+| Article could not be loaded | FAILED_PRECONDITION | 503 | max-age=0, no-cache, no-store, must-revalidate
+| Article loading timeout | FAILED_PRECONDITION | 503 | max-age=0, no-cache, no-store, must-revalidate
```
Comment (moritzzimmer) [inline docs/service_status_and_errors.md:18]: | scenario | description | gRPC status | gRPC error payload | HTTP Status | cacheable | 
|:---|:---|:---:|:---|:---:|:---:|
| found | article was found in the datastore and is published according to it's [Metadata](https://github.com/stroeer/tapir/blob/d99e155003a5f81a4f300d558d1fbc736dbda511/stroeer/core/v1/article.proto#L87)| OK | none | 200 | yes
| invalid id | article id is invalid | INVALID_ARGUMENT | none | 400 | yes
| not found | article was not found in the datastore | NOT_FOUND | [google.rpc.ResourceInfo](https://github.com/googleapis/googleapis/blob/3e9b231b098d514c9a51ba8c59e416f45837b195/google/rpc/error_details.proto#L198) | 404 | yes
| not published yet | article was found in the datastore but it's `Metadata` [start_time](https://github.com/stroeer/tapir/blob/d99e155003a5f81a4f300d558d1fbc736dbda511/stroeer/core/v1/article.proto#L93) is not reached, yet | NOT_FOUND | [google.rpc.ResourceInfo](https://github.com/googleapis/googleapis/blob/3e9b231b098d514c9a51ba8c59e416f45837b195/google/rpc/error_details.proto#L198) with info about the `start_time` | 404 | yes
| expired | article was found in the datastore but is expired according to it's `Metadata` [end_time](https://github.com/stroeer/tapir/blob/d99e155003a5f81a4f300d558d1fbc736dbda511/stroeer/core/v1/article.proto#L97) | NOT_FOUND | [google.rpc.ResourceInfo](https://github.com/googleapis/googleapis/blob/3e9b231b098d514c9a51ba8c59e416f45837b195/google/rpc/error_details.proto#L198) with info about the `end_time` | 404 | yes
| deleted/archived (TODO: clarify requirements with SEO) | article was found in the datastore but it's [state](https://github.com/stroeer/tapir/blob/d99e155003a5f81a4f300d558d1fbc736dbda511/stroeer/core/v1/article.proto#L89) is marked as deleted/archived | FAILED_PRECONDITION | [google.rpc.PreconditionFailure](https://github.com/googleapis/googleapis/blob/3e9b231b098d514c9a51ba8c59e416f45837b195/google/rpc/error_details.proto#L143) with info about the [state](https://github.com/stroeer/tapir/blob/d99e155003a5f81a4f300d558d1fbc736dbda511/stroeer/core/v1/article.proto#L89) | 410 | yes
| internal | internal error processing the article | INTERNAL | none | 500 | no
| timeout | timeout loading and processing the article | DEADLINE_EXCEEDED | none | 504 | no

File: docs/service_status_and_errors.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+## Section Page Service Status
+
+| Error scenario | GRPC status | HTTP Status | Cache Headers | Description |
+| --- |:---:| ---:| ---:| --- |
+| Section path is empty| INVALID_ARGUMENT | 400 | max-age=60
+| Section path invalid | INVALID_ARGUMENT | 400 | max-age=60
+| Section path unknown | FAILED_PRECONDITION | 404 | max-age=60
```
Comment (moritzzimmer) [inline docs/service_status_and_errors.md:26]: Maybe it's good enough to return with a cacheable `NOT_FOUND`/`404` for all those cases, gRPC API can include details in a `google.rpc.ResourceInfo`

File: docs/service_status_and_errors.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+| Stage data not found | FAILED_PRECONDITION | 404 | max-age=60
+| Stage data is empty | FAILED_PRECONDITION | 404 | max-age=60
+| Stage datasource is not available | FAILED_PRECONDITION | 503 | max-age=0, no-cache, no-store, must-revalidate
+| Stage loading timeout | FAILED_PRECONDITION | 503 | max-age=0, no-cache, no-store, must-revalidate
```
Comment (moritzzimmer) [inline docs/service_status_and_errors.md:30]: `DEADLINE_EXCEEDED`/`504` like proposed for article page service

File: docs/service_status_and_errors.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+## Section Page Service Status
+
+| Error scenario | GRPC status | HTTP Status | Cache Headers | Description |
+| --- |:---:| ---:| ---:| --- |
+| Section path is empty| INVALID_ARGUMENT | 400 | max-age=60
+| Section path invalid | INVALID_ARGUMENT | 400 | max-age=60
+| Section path unknown | FAILED_PRECONDITION | 404 | max-age=60
+| Stage data not found | FAILED_PRECONDITION | 404 | max-age=60
+| Stage data is empty | FAILED_PRECONDITION | 404 | max-age=60
+| Stage datasource is not available | FAILED_PRECONDITION | 503 | max-age=0, no-cache, no-store, must-revalidate
```
Comment (moritzzimmer) [inline docs/service_status_and_errors.md:29]: non-cacheable `INTERNAL`/`500`

File: docs/service_status_and_errors.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+## Section Page Service Status
+
+| Error scenario | GRPC status | HTTP Status | Cache Headers | Description |
+| --- |:---:| ---:| ---:| --- |
+| Section path is empty| INVALID_ARGUMENT | 400 | max-age=60
+| Section path invalid | INVALID_ARGUMENT | 400 | max-age=60
+| Section path unknown | FAILED_PRECONDITION | 404 | max-age=60
+| Stage data not found | FAILED_PRECONDITION | 404 | max-age=60
+| Stage data is empty | FAILED_PRECONDITION | 404 | max-age=60
```
Comment (moritzzimmer) [inline docs/service_status_and_errors.md:28]: IMO we should clarify with product/seo how we should handle missing stages/stage_items and use the tapir lingo here

File: docs/service_status_and_errors.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+| Article found | OK | 200 | max-age=60
+| Article not found | NOT_FOUND | 404 | max-age=60
+| Article not published yet | NOT_FOUND | 404 | max-age=60
+| Article expired | NOT_FOUND | 410 | max-age=60
+| Article could not be loaded | FAILED_PRECONDITION | 503 | max-age=0, no-cache, no-store, must-revalidate
+| Article loading timeout | FAILED_PRECONDITION | 503 | max-age=0, no-cache, no-store, must-revalidate
+
+## Section Page Service Status
+
+| Error scenario | GRPC status | HTTP Status | Cache Headers | Description |
```
Comment (moritzzimmer) [inline docs/service_status_and_errors.md:22]: `OK` path is missing

File: docs/service_status_and_errors.md (hunk 6 above)
Comment (FalkoBasner) [reply docs/service_status_and_errors.md]: For now I removed this detailed stage handling

File: docs/service_status_and_errors.md (hunk 3 above)
Comment (FalkoBasner) [reply docs/service_status_and_errors.md]: For 400 we could use [google.rpc.BadRequest](https://github.com/googleapis/googleapis/blob/3e9b231b098d514c9a51ba8c59e416f45837b195/google/rpc/error_details.proto#L169).

**Comment by FalkoBasner** [comment]
1. Moritz mentioned [google.rpc.ResourceInfo](https://github.com/googleapis/googleapis/blob/3e9b231b098d514c9a51ba8c59e416f45837b195/google/rpc/error_details.proto#L198). This proto defines error messages for e.g. [google.rpc.BadRequest](https://github.com/googleapis/googleapis/blob/3e9b231b098d514c9a51ba8c59e416f45837b195/google/rpc/error_details.proto#L169) as well. Should we use these messages as payload for errors ? 

2. Incomplete Section Data scenarios needs to be defined

3. Redirect scenarios needs to be defined

**Review by moritzzimmer — APPROVED** [review]

**Pull request merged**